File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> path;
    int sum = 0;
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        backtracking(candidates, target, 0);
        return res;
    }

    void backtracking(vector<int>& candidates, int target, int start) {
        if (sum == target) {
            vector<int> copy = path;
            res.push_back(copy);
            return;
        }

        for (int i = start; i < candidates.size(); i++) {
            int val = candidates[i];
            if (sum + val > target) {
                break;
            }

            if (i != start && val == candidates[i - 1]) {
                continue;
            }

            path.push_back(val);
            sum += val;
            backtracking(candidates, target, i + 1);
            sum -= val;
            path.pop_back();

        }
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def __init__(self):
        self.path = []
        self.res = []

    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        self.backtracking(candidates, target, 0, 0)
        return self.res
    
    def backtracking(self, candidates: List[int], target: int, start: int, currSum: int):
        if currSum == target:
            self.res.append(copy.copy(self.path))
            return
        
        for i in range(start, len(candidates)):
            if i > start and candidates[i] == candidates[i - 1]:
                continue
            
            if currSum + candidates[i] > target:
                break
            
            self.path.append(candidates[i])
            self.backtracking(candidates, target, i + 1, currSum + candidates[i])
            self.path.pop()
